debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsStreamProducerIT.java

```diff
@@ -78,6 +80,10 @@ public void before() throws Exception {
         // ensure the slot is deleted for each test
         try (PostgresConnection conn = TestHelper.create()) {
             conn.dropReplicationSlot(ReplicationConnection.Builder.DEFAULT_SLOT_NAME);
+            // DBZ-766 should only drop publication when using PGOUTPUT
```

Comment:
And only in case we created it ourselves, right?

Reply:
Actually the only place in the code where it gets dropped is that test.  At runtime even if we create the publication ourselves I'm not dropping it, which may be incorrect in hindsight.  It would seem if we create it ourselves we should remove it when the connector stops.

Reply:
I'm not so sure; doesn't equate to deleting the replication slot and thus potentially losing events during a downtime? Probably it makes most sense to handle it similarly to the publication slot itself: create it of not existing and not remove it.

Reply:
So I added support in `PostgresReplicationConnection` to drop the publication if and only if it did not exist and we created it when the connection was initialized.  What I am facing is some odd PostgreSQL behavior that I cannot yet explain or understand.

The exception itself is the following:
```
org.postgresql.util.PSQLException: ERROR: publication "dbz_publication" does not exist
  Where: slot "debezium", output plugin "pgoutput", in the change callback, associated LSN 0/6DD1AF8
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2440)
	at org.postgresql.core.v3.QueryExecutorImpl.processCopyResults(QueryExecutorImpl.java:1116)
	at org.postgresql.core.v3.QueryExecutorImpl.readFromCopy(QueryExecutorImpl.java:1035)
	at org.postgresql.core.v3.CopyDualImpl.readFromCopy(CopyDualImpl.java:41)
	at org.postgresql.core.v3.replication.V3PGReplicationStream.receiveNextData(V3PGReplicationStream.java:155)
	at org.postgresql.core.v3.replication.V3PGReplicationStream.readInternal(V3PGReplicationStream.java:124)
	at org.postgresql.core.v3.replication.V3PGReplicationStream.readPending(V3PGReplicationStream.java:78)
	at io.debezium.connector.postgresql.connection.PostgresReplicationConnection$1.readPending(PostgresReplicationConnection.java:326)
	at io.debezium.connector.postgresql.RecordsStreamProducer.streamChanges(RecordsStreamProducer.java:162)
	at io.debezium.connector.postgresql.RecordsStreamProducer.lambda$start$0(RecordsStreamProducer.java:144)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:514)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.base/java.lang.Thread.run(Thread.java:844)
```
So far this only happens under this scenario AFAICT:

1. Start connector with `DROP_SLOT_ON_STOP=false`
    a. Creates `dbz_publication` since it does not exist.
    b. Creates replication slot `debezium`
    c. Snapshots/Streams changes
2. Stop the connector
    a. Drops the `dbz_publication` since we created it in (1)
3. Perform some SQL operations while connector is down
4. Restart the connector, this time with `DROP_SLOT_ON_STOP=true`
    a. Creates `dbz_publication` since it does not exist.
    b. Reuses existing replication slot `debezium` that is not active.
    c. Streams changes <- failure happens here calling `PGReplicationStream#readPending`

The only solution I've been able to explore so far is to only drop the publication if and only if the replication slot is also dropped.  I'm wondering if @jpechane isn't right and we likely should simply delegate publication management to the user entirely.

Reply:
@gunnarmorling Apparently we replied at near the same time and I missed your reply above; yah in that case everything I mentioned above becomes moot then so disregard it then.

Reply:
>  I'm wondering if @jpechane isn't right and we likely should simply delegate publication management to the user entirely.

Yes, that'd be totally fine for the first iteration for sure. We then can improve it to create the publication if a) it doesn't exist and b) we have the right permissions to do so.

Reply:
Just be careful wit too smart fallbacks - https://issues.jboss.org/browse/DBZ-1399